File: apps/frontend/lib/src/features/maintenance_ticket/domain/entities/sla_configuration_entity.dart
import 'package:equatable/equatable.dart';

import 'maintenance_ticket_entity.dart';

class SlaConfigurationEntity extends Equatable {
  const SlaConfigurationEntity({
    required this.id,
    required this.companyId,
    required this.name,
    this.description,
    required this.priority,
    required this.firstResponseTimeMinutes,
    required this.acknowledgementTimeMinutes,
    required this.resolutionTimeMinutes,
    this.escalationLevel1Minutes,
    this.escalationLevel2Minutes,
    this.escalationLevel3Minutes,
    this.applyBusinessHours = true,
    this.businessStartTime,
    this.businessEndTime,
    this.workingDays,
    this.excludeHolidays = true,
    this.isActive = true,
    this.createdById,
    required this.createdAt,
    required this.updatedAt,
  });

  final String id;
  final String companyId;
  final String name;
  final String? description;
  final TicketPriority priority;
  final int firstResponseTimeMinutes;
  final int acknowledgementTimeMinutes;
  final int resolutionTimeMinutes;
  final int? escalationLevel1Minutes;
  final int? escalationLevel2Minutes;
  final int? escalationLevel3Minutes;
  final bool applyBusinessHours;
  final String? businessStartTime; // HH:MM format
  final String? businessEndTime; // HH:MM format
  final String? workingDays; // Comma-separated: "1,2,3,4,5" (Mon-Fri)
  final bool excludeHolidays;
  final bool isActive;
  final String? createdById;
  final DateTime createdAt;
  final DateTime updatedAt;

  String get priorityDisplayName => priority.displayName;

  String get workingDaysDisplay {
    if (workingDays == null || workingDays!.isEmpty) {
      return 'All Days';
    }
    final days = workingDays!.split(',');
    final dayNames = ['Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun'];
    return days
        .map((d) {
          final index = int.tryParse(d.trim());
          if (index != null && index >= 1 && index <= 7) {
            return dayNames[index - 1];
          }
          return '';
        })
        .where((d) => d.isNotEmpty)
        .join(', ');
  }

  @override
  List<Object?> get props => [
        id,
        companyId,
        name,
        description,
        priority,
        firstResponseTimeMinutes,
        acknowledgementTimeMinutes,
        resolutionTimeMinutes,
        escalationLevel1Minutes,
        escalationLevel2Minutes,
        escalationLevel3Minutes,
        applyBusinessHours,
        businessStartTime,
        businessEndTime,
        workingDays,
        excludeHolidays,
        isActive,
        createdById,
        createdAt,
        updatedAt,
      ];
}

File: apps/backend/src/modules/maintenance-ticket/entities/sla-configuration.entity.ts
import {
  Entity,
  Column,
  Index,
  ManyToOne,
  JoinColumn,
} from 'typeorm';
import { TenantBaseEntity } from '../../../shared/database/tenant-base.entity';
import { TicketPriority } from '../enums/ticket-priority.enum';

@Entity('sla_configurations')
@Index(['companyId', 'priority'], { unique: true })
@Index(['companyId', 'isActive'])
export class SlaConfiguration extends TenantBaseEntity {
  @Column({ type: 'varchar', length: 100 })
  name!: string;

  @Column({ type: 'text', nullable: true })
  description?: string;

  @Column({
    type: 'enum',
    enum: TicketPriority,
  })
  priority!: TicketPriority;

  // First response time in minutes
  @Column({ type: 'int', name: 'first_response_time_minutes' })
  firstResponseTimeMinutes!: number;

  // Acknowledgement time in minutes
  @Column({ type: 'int', name: 'acknowledgement_time_minutes' })
  acknowledgementTimeMinutes!: number;

  // Start work time in minutes (deadline to change status from ASSIGNED to IN_PROGRESS)
  @Column({ type: 'int', name: 'start_work_time_minutes', nullable: true })
  startWorkTimeMinutes?: number;

  // Resolution time in minutes
  @Column({ type: 'int', name: 'resolution_time_minutes' })
  resolutionTimeMinutes!: number;

  // Escalation thresholds
  @Column({ type: 'int', name: 'escalation_level_1_minutes', nullable: true })
  escalationLevel1Minutes?: number;

  @Column({ type: 'int', name: 'escalation_level_2_minutes', nullable: true })
  escalationLevel2Minutes?: number;

  @Column({ type: 'int', name: 'escalation_level_3_minutes', nullable: true })
  escalationLevel3Minutes?: number;

  // Business hours settings
  @Column({ type: 'boolean', default: true, name: 'apply_business_hours' })
  applyBusinessHours!: boolean;

  @Column({ type: 'time', nullable: true, name: 'business_start_time' })
  businessStartTime?: string;

  @Column({ type: 'time', nullable: true, name: 'business_end_time' })
  businessEndTime?: string;

  // Working days (comma-separated: 1,2,3,4,5 for Mon-Fri)
  @Column({ type: 'varchar', length: 20, nullable: true, name: 'working_days' })
  workingDays?: string;

  // Exclude holidays
  @Column({ type: 'boolean', default: true, name: 'exclude_holidays' })
  excludeHolidays!: boolean;

  @Column({ type: 'boolean', default: true, name: 'is_active' })
  isActive!: boolean;

  @Column({ type: 'uuid', nullable: true, name: 'created_by_id' })
  createdById?: string;
}

